src/Raven.Server/Documents/DocumentDatabase.cs

```diff
@@ -806,13 +811,18 @@ public void StateChanged(long index)
             }
             catch (Exception e)
             {
+                DatabaseDisabledException throwShutDown = null;
+
+                if (_databaseShutdown.IsCancellationRequested && e is DatabaseDisabledException == false)
+                    e = throwShutDown = CreateDatabaseShutdownException(e);
+
                 RachisLogIndexNotifications.NotifyListenersAbout(index, e);
 
                 if (_logger.IsInfoEnabled)
                     _logger.Info($"Got exception during StateChanged({index}).", e);
 
-                if (_databaseShutdown.IsCancellationRequested)
-                    ThrowDatabaseShutdown(e);
+                if (throwShutDown != null)
+                    throw throwShutDown;
```

Comment:
We want the "throw" code to be encapsulated in a separate function in order to reduce generated function code size and complexity

Reply:
I know but:
- we have `throw` here anyway as we rethrow the exception
- I don't think that it's hot path because `StateChanged` is called really occasionally